sklearn/metrics/cluster/_unsupervised.py

```diff
@@ -131,15 +132,20 @@ def _silhouette_reduce(D_chunk, start, labels, label_freqs):
     label_freqs : array-like
         Distribution of cluster labels in ``labels``.
     """
+    n_chunk_samples = D_chunk.shape[0]
     # accumulate distances from each sample to each cluster
-    clust_dists = np.zeros((len(D_chunk), len(label_freqs)),
+    clust_dists = np.zeros((n_chunk_samples, len(label_freqs)),
                            dtype=D_chunk.dtype)
-    for i in range(len(D_chunk)):
-        clust_dists[i] += np.bincount(labels, weights=D_chunk[i],
+    for i in range(n_chunk_samples):
+        sample_weights = (
+            D_chunk.getrow(i).toarray().squeeze()
+            if issparse(D_chunk) else D_chunk[i])
```

Comment:
This would make the sparse matrix dense, row by row. What does the memory usage look like for different number of samples?

Reply:
I didn't benchmark the memory usage. But, converting the sparse matrix to dense would definitely spike it up for a large number of samples. I have updated the logic to avoid doing this